drivers: spi: spi_context: exclude gpio code if no gpio cs

maass-hamburg opened the pull request:
exclude gpio code if no gpio cs are used
for that spi driver. Because all of it is in a header and all functions and definitions are inline, we can optimize code out during build

maass-hamburg committed:
drivers: spi: spi_context: exclude gpio code if no gpio cs

exclude gpio code if no gpio cs are used
for that spi driver.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

tbursztyka reviewed (commented):
please rebase it.

Seems ok to me, though perhaps it could reduce the amount of ifdefs by doing so: 
```
#ifndef  DT_SPI_CTX_HAS_NO_CS_GPIOS
... declare all the gpio cs functions asdoing nothing ...
# else
... the current function doing gpio cs stuff ...
#endif
```

pdgendt commented on drivers/spi/spi_context.h:
```diff
@@ -292,6 +302,13 @@ static inline void spi_context_complete(struct spi_context *ctx,
 }
 
+#ifdef DT_SPI_CTX_HAS_NO_CS_GPIOS
+#define spi_context_cs_configure_all(...)
+#define spi_context_cs_get_all(...)
+#define spi_context_cs_put_all(...)
```
These functions have `int` return types. Could be defined as `0`, not sure if this could cause other issues?

maass-hamburg committed:
drivers: spi: remove spi_cs_is_gpio checks

remove spi_cs_is_gpio checks before
spi_context_cs_control, as it is also done
inside and we don't need to check two times.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

ubieda commented:
Coming to report this PR breaks NXP's LPSPI driver with native CS. I wonder if other platforms are also affected as well.

Has this been validated on platforms other than the ones that have been changed (e.g: ESP32, STM32 and Numaker)? If not, given this PR appears to be about optimizing code vs adding functionality, we may want to consider reverting for now.

maass-hamburg commented:
@ubieda The driver is split into multiple files. In `drivers/spi/spi_nxp_lpspi/spi_nxp_lpspi.c` and `drivers/spi/spi_nxp_lpspi/spi_nxp_lpspi_dma.c` `DT_DRV_COMPAT` is defined and in `drivers/spi/spi_nxp_lpspi/spi_nxp_lpspi_common.c` it isn't
